File: cleanclaw/projectmap/classifier.py
from pathlib import Path

_HEURISTICS: dict[str, set[str]] = {
    "backend": {
        "backend", "api", "server", "controllers", "services", "repositories",
        "handlers", "domain", "data", "models", "entities", "infrastructure",
        "persistence", "database", "db", "dal", "bll", "business", "core",
        "workers", "jobs", "tasks", "queues", "events", "notifications",
    },
    "frontend": {
        "frontend", "client", "ui", "components", "pages", "views", "screens",
        "app", "assets", "styles", "scss", "css", "layouts", "templates",
        "widgets", "hooks", "store", "reducers", "actions", "selectors",
    },
    "mediator": {
        "mediator", "gateway", "middleware", "hubs", "proxy", "bridge",
        "adapters", "connectors", "integrations", "transforms", "mappers",
        "normalizers", "validators", "filters", "interceptors",
    },
}


def classify_file(file_path: str, layer_map: dict[str, str] | None = None,
                  extra_keywords: dict[str, list[str]] | None = None) -> str:
    """Return 'backend', 'frontend', 'mediator', or 'misc' for the given path.

    layer_map:     exact prefix overrides from config  e.g. {"src/Shared": "backend"}
    extra_keywords: additional keywords per layer from config
                    e.g. {"backend": ["myservice"], "frontend": ["portal"]}
    """
    if layer_map:
        for prefix, layer in layer_map.items():
            if file_path.startswith(prefix):
                return layer

    heuristics = _HEURISTICS
    if extra_keywords:
        heuristics = {
            layer: keywords | set(extra_keywords.get(layer, []))
            for layer, keywords in _HEURISTICS.items()
        }

    parts = {p.lower() for p in Path(file_path).parts}
    for layer, keywords in heuristics.items():
        if parts & keywords:
            return layer

    return "misc"

File: cleanclaw/projectmap/classifier.ts
import path from "path";

const _HEURISTICS: Record<string, Set<string>> = {
  backend: new Set([
    "backend", "api", "server", "services", "repositories",
    "handlers", "domain", "data", "models", "entities", "infrastructure",
    "persistence", "database", "db", "dal", "bll", "business", "core",
    "workers", "jobs", "tasks", "queues", "events", "notifications",
  ]),
  frontend: new Set([
    "frontend", "client", "ui", "components", "pages", "views", "screens",
    "app", "assets", "styles", "scss", "css", "layouts", "templates",
    "widgets", "hooks", "store", "reducers", "actions", "selectors",
  ]),
  mediator: new Set([
    "mediator", "gateway", "middleware", "hubs", "proxy", "bridge", "controllers",
    "adapters", "connectors", "integrations", "transforms", "mappers",
    "normalizers", "validators", "filters", "interceptors",
  ]),
};

export function classifyFile(
  filePath: string,
  layerMap?: Record<string, string> | null,
  extraKeywords?: Record<string, string[]> | null
): string {
  if (layerMap) {
    for (const [prefix, layer] of Object.entries(layerMap)) {
      if (filePath.startsWith(prefix)) {
        return layer;
      }
    }
  }

  let heuristics = _HEURISTICS;
  if (extraKeywords) {
    heuristics = {} as Record<string, Set<string>>;
    for (const [layer, keywords] of Object.entries(_HEURISTICS)) {
      const extra = extraKeywords[layer] ?? [];
      heuristics[layer] = new Set([...keywords, ...extra]);
    }
  }

  const parts = new Set(
    filePath.split(/[\\/]/).map((p) => p.toLowerCase())
  );

  for (const [layer, keywords] of Object.entries(heuristics)) {
    for (const kw of keywords) {
      if (parts.has(kw)) {
        return layer;
      }
    }
  }

  return "misc";
}
